### PR TITLE
Add support for hash template syntax

### PR DESCRIPTION
**Pull request by supernintendo**
This PR adds hash syntax for template IDs so instead of this:

```heex
<Menu>
  <Text template={:label}>
    Open Me
  </Text>
</Menu>
```

You can do this:

```heex
<Menu>
  <Text #label>
    Open Me
  </Text>
</Menu>
```

This is a follow-up to the conversation [in this thread](https://github.com/liveview-native/liveview-client-swiftui/pull/787#discussion_r1170453190) and requires the upstream changes to phoenix_live_view [proposed here](https://github.com/phoenixframework/phoenix_live_view/pull/2636).

Closes https://github.com/liveview-native/live_view_native/issues/15

### CODE REVIEW

**Comment by josevalim** [comment]
Just so I understand things better, what could be instead of `Text("Open Me")` in the "label" bits? I guess it could be anything?

```
Menu {
  Button("Option 1") {}
  Button("Option 2") {}
  Button("Option 3") {}
} label: {
  Text("Open Me")
}
```

**Comment by josevalim** [comment]
Btw, thanks for the PR. At first, I think the proposed hook is too powerful. You can change the template in any way, perhaps in ways that are then incompatible with the engine. I would also prefer if we don't change the attribute syntax. The goal is to keep those templates "similar". We already allow differences in the tag name, so maybe we could build on that?

```
<Menu>
  <label:Text>
    Open Me
  </label:Text>
  <content:Group>
    <Button>Option 1</Button>
    <Button>Option 2</Button>
    <Button>Option 3</Button>
  </content:Group>
</Menu>

File: lib/live_view_native/tag_engine.ex [hunk 1]
```diff
@@ -17,5 +18,30 @@ defmodule LiveViewNative.TagEngine do
...
+
+  defp expand_template_attribute({attr_name, nil, metadata} = attr) do
+    if String.starts_with?(attr_name, "#") do
+      template_name = String.replace_leading(attr_name, "#", ":")
+
+      {"template", {:expr, template_name, metadata}, metadata}
+    else
+      attr
+    end
+  end
```
Comment (carson-katri) [inline lib/live_view_native/tag_engine.ex:44]: This is a bit simpler and I think has the same behavior:
```suggestion
  defp expand_template_attribute({"#" <> rest, nil, metadata} = attr) do
    {"template", {:expr, ":" <> rest, metadata}, metadata}
  end
```

**Comment by supernintendo** [comment]
Hi @josevalim, thanks for your feedback 🙂 

> Just so I understand things better, what could be instead of `Text("Open Me")` in the "label" bits? I guess it could be anything?

Yes, that's correct.

> Btw, thanks for the PR. At first, I think the proposed hook is too powerful. You can change the template in any way, perhaps in ways that are then incompatible with the engine. I would also prefer if we don't change the attribute syntax.

Consistency with standard LiveView is definitely a concern and one of the design goals of LiveView Native. We want to make it as easy as possible for Elixir developers who are familiar with Phoenix and LiveView to build mobile and desktop apps without having to learn our own syntax. But I do think there are some cases where decorating the normal HEEx syntax with our own could be helpful, especially for surfacing functionality that might be difficult to represent with just HTML as there is no comparable functionality on the web (i.e. SwiftUI modifiers with nested content). 

My proposed hook being too powerful is valid and if you think introducing it will encourage bad practices then I'm okay with scrapping it in favor of another solution. Would a callback that only allows you to transform attributes within HEEx templates at compile-time rather than giving control over all tokens in the AST be sufficient? I might also be able to abstract it further to prohibit modifying the token structure so that only the parts we need to change are exposed.

> The goal is to keep those templates "similar". We already allow differences in the tag name, so maybe we could build on that?

We did originally use a template ID format like what you're proposing but moved away from that for a few reasons. Concatenating the element and ID like that is syntactically clunky and diminished the ergonomics of writing templates for LVN, so we abandoned it in favor of just using a `template` attribute. It also required some additional parsing on the side of the client which we're trying to avoid (see [this thread](https://github.com/liveview-native/liveview-client-swiftui/pull/787#discussion_r1170453190)). I personally find the hash syntax easier to understand and more appealing to write, and we have had some interest in it like [this comment](https://github.com/liveview-native/liveview-client-swiftui/pull/787#discussion_r1178187537). However, I'm not sure it's a make or break feature unless @bcardarella thinks so?

**Comment by josevalim** [comment]
> We did originally use a template ID format like what you're proposing but moved away from that for a few reasons. Concatenating the element and ID like that is syntactically clunky and diminished the ergonomics of writing templates for LVN, so we abandoned it in favor of just using a template attribute. It also required some additional parsing on the side of the client which we're trying to avoid (see https://github.com/liveview-native/liveview-client-swiftui/pull/787#discussion_r1170453190).

What if we improve TagHandler such that `<label:Text>` becomes `<Text template={:label}>`? This way you don't need to parse on the client, you keep the higher-level template API, and we don't need to add syntax for custom attributes. We would, however, need to improve TagHandler to allow initial attributes to be returned when parsing a tag.

Thoughts?

**Comment by supernintendo** [comment]
> > We did originally use a template ID format like what you're proposing but moved away from that for a few reasons. Concatenating the element and ID like that is syntactically clunky and diminished the ergonomics of writing templates for LVN, so we abandoned it in favor of just using a template attribute. It also required some additional parsing on the side of the client which we're trying to avoid (see [liveview-native/liveview-client-swiftui#787 (comment)](https://github.com/liveview-native/liveview-client-swiftui/pull/787#discussion_r1170453190)).
> 
> What if we improve TagHandler such that `<label:Text>` becomes `<Text template={:label}>`? This way you don't need to parse on the client, you keep the higher-level template API, and we don't need to add syntax for custom attributes. We would, however, need to improve TagHandler to allow initial attributes to be returned when parsing a tag.
> 
> Thoughts?

Would that allow us to support the hash syntax (i.e `<Text #label>) instead? We moved away from the `<label:Text>` syntax for the reasons I described; if this is something we absolutely can't support it would be disappointing, but my thoughts is that we would just stick with the `template` attribute for template IDs as we're already doing.

**Comment by supernintendo** [comment]
@josevalim Also to be clear, the changes I'm proposing only affect our own custom templates for native clients like iOS and Android. Normal HEEx templates are unaffected.

**Comment by josevalim** [comment]
> We moved away from the <label:Text> syntax for the reasons I described

No, it would not allow `<Text #label>`. But the proposed mechanism should address the drawbacks you described in your original implementation of `<label:Text>`. Is my proposal not enough?

> Normal HEEx templates are unaffected.

Yup, all clear.

**Comment by supernintendo** [comment]
@josevalim That does solve the issue of having to parse the template ID from the native client but it doesn't address the aesthetics of `<label:Text>` which everyone on the LVN team previously agreed is undesirable. We can open that discussion back up but my sense is that if hash syntax can't be supported, we would stick with template IDs as they are currently:

```
<Menu>
  <Text template={:label}>Label</Text>
  <Group template={:content}>
    <Button>Option 1</Button>
    <Button>Option 2</Button>
    <Button>Option 3</Button>
  </Group>
</Menu>
```

**Comment by josevalim** [comment]
> We can open that discussion back up but my sense is that if hash syntax can't be supported, we would stick with template IDs as they are currently

I think one of the two are the safest bets for now, thanks!

**Comment by supernintendo** [comment]
@josevalim Thank you so much for giving this more consideration! I see your points and am curious what the rest of the team working on the SwiftUI client thinks (@carson-katri @shadowfacts @dginsburg @bcardarella). Here are my thoughts:

> label:Text more closely reflects how the syntax looks like in Swift UI (it has the same label: bits)

I suppose this is true to a degree, although I think the resemblance seems a bit more superficial to me since the `label:` in SwiftUI is just the syntax for defining parameters of `Menu` whereas here we're using it more as a prefix of the `Text` element to indicate which parameter it maps to. The more problematic aspect of this syntax, though, is demonstrated by closing tags:

```heex
<Menu>
  <label:Text>
    Some text
  </label:Text>
</Menu>
```

From my experience using this syntax, especially with more complex LiveViews, having to include the template ID in the closing tag is cumbersome and makes for some pretty ugly templates with all the duplication. These template IDs are not always static either; some SwiftUI modifiers allow you to define custom IDs, like [here](https://github.com/liveview-native/scratchboard/blob/main/lib/scratchboard_web/components/modifiers/layout_fundamentals.ex#L12) which makes the syntax have an even worse feel when introducing underscores.

> #label in place of an attribute is foreign in HTML. The closest thing it resembles to me is the ID notation in CSS. Given templating languages like HAML, I wouldn't be surprised if people read <Text #label> as <Text id="label">

True but isn't HEEx already full of non-HTML like syntax (EEx interpolation, curly brackets, slots, function components, etc.)? I don't understand why we would draw the line here, especially since the platform we're targeting in this case (SwiftUI) doesn't even render HTML. Also, I think deviating from conventional syntax is fine if it improves the ergonomics of LiveView Native. The problem we're trying to solve is makingit easy to use and thus, making templates easy to write - it is _not_ consistency with HTML.

Also, there is precedent for this syntax in other popular frameworks. Vue.js uses hash syntax for an almost identical use case in slots:

```vue
<BaseLayout>
  <template #header>
    <h1>Here might be a page title</h1>
  </template>

  <template #default>
    <p>A paragraph for the main content.</p>
    <p>And another one.</p>
  </template>

  <template #footer>
    <p>Here's some contact info</p>
  </template>
</BaseLayout>
```

Your point about people conflating `#label` with `id="label"` is valid but I'm wondering why you think this is detrimental as that's essentially the purpose of the template ID? It's a unique ID to reference that element from a modifier function (the equivalent of CSS for SwiftUI).

**Comment by josevalim** [comment]
> True but isn't HEEx already full of non-HTML like syntax (EEx interpolation, curly brackets, slots, function components, etc.)? I don't understand why we would draw the line here

Yes, but adding #label would start causing larger differences between different "HEEx"-engines. And I would imagine that, even if LVN does not render HTML, it would be best to deviate from LiveView's own engine as little as possible?

Plus, if you look at HEEx design, we put most of the notation in the tag names. While we do support attribute interpolation, function components, slots, etc are all variations of the tag. That's why we chose something like `<:slot>` for slots rather than `<template #slot>`. The idea is that the tag names may be foreign but the attributes still feel fairly familiar. Adding `<Text #foo>` would be a big departure from those earlier decisions.

**Comment by supernintendo** [comment]
@josevalim Fair points, especially the concern over diverging HEEx engines out in the wild. I think after discussing this internally, we've decided to close this issue and stick with what we have for now (simply representing the template IDs as an attribute, i.e. `template={:label}`). This is already supported, doesn't involve learning any sort of custom syntax and has an advantage (albeit niche) that all of the other proposals don't in that it allows passing an assign as a template ID (i.e. `template={@template_id}`). If we want to continue the discussion around template ID syntax, we can do so as part of another issue.

Thanks for the thoughtful discussion and feedback José, its really helpful in determining the best path forward for LiveView Native so it is always greatly appreciated 😊

**Comment by josevalim** [comment]
:+1: for focusing on the foundation. Syntax sugar can always come later!

**Pull request closed without merging**